src/registrar.js

```diff
@@ -0,0 +1,165 @@
+'use strict'
+
+const assert = require('assert')
+const debug = require('debug')
+const log = debug('libp2p:peer-store')
+log.error = debug('libp2p:peer-store:error')
+const errCode = require('err-code')
+
+const { Connection } = require('libp2p-interfaces/src/connection')
+const PeerInfo = require('peer-info')
+const MulticodecTopology = require('./connection-manager/topology')
+
+/**
+ * Responsible for notifying registered protocols of events in the network.
+ */
+class Registrar {
+  /**
+   * @param {PeerStore} peerStore
+   * @constructor
+   */
+  constructor (peerStore) {
+    this.peerStore = peerStore
+
+    /**
+     * Map of connections per peer
+     * TODO: this should be handled by connectionManager
+     * @type {Map<string, conn>}
+     */
+    this.connections = new Map()
+
+    /**
+     * Map of topologies per multicodec
+     *
+     * @type {Map<string, object>}
+     */
+    this.multicodecTopologies = new Map()
+
+    this._handle = undefined
+  }
+
+  get handle () {
+    return this._handle
+  }
+
+  set handle (handle) {
+    this._handle = handle
+  }
+
+  /**
+   * Add a new connected peer to the record
+   * TODO: this should live in the ConnectionManager
+   * @param {PeerInfo} peerInfo
+   * @param {Connection} conn
+   * @returns {void}
+   */
+  onConnect (peerInfo, conn) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+    assert(Connection.isConnection(conn), 'conn must be an instance of interface-connection')
+
+    this.connections.set(peerInfo.id.toB58String(), conn)
+  }
+
+  /**
+   * Remove a disconnected peer from the record
+   * TODO: this should live in the ConnectionManager
+   * @param {PeerInfo} peerInfo
+   * @param {Error} [error]
+   * @returns {void}
+   */
+  onDisconnect (peerInfo, error) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+
+    for (const [, topology] of this.multicodecTopologies) {
+      topology.disconnect(peerInfo, error)
+    }
+
+    this.connections.delete(peerInfo.id.toB58String())
+  }
+
+  /**
+   * Get a connection with a peer.
+   * @param {PeerInfo} peerInfo
+   * @returns {Connection}
+   */
+  getPeerConnection (peerInfo) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peerInfo must be an instance of peer-info')
+
+    return this.connections.get(peerInfo.id.toB58String())
+  }
+
+  /**
+   * Register handlers for a set of multicodecs given
+   * @param {Array<string>|string} multicodecs
+   * @param {object} handlers
+   * @param {function} handlers.onConnect
+   * @param {function} handlers.onDisconnect
+   * @param {object} topology properties for topology
+   * @return {string} registrar identifier
+   */
+  register (multicodecs, handlers, topology = {}) {
```

Comment:
`topology` should be required. The registrar shouldn't create them, it should just track and inform them. With that the `handlers` shouldn't need to be passed, because they're part of the topology.

`multicodecs` shouldn't be needed and should exist solely in the `topology`

Reply:
In this scenario, I am also using `multicodecs` to try to connect to the already connected peers using that multicodec. If I don't use them here, I will need to try `onConnect` to all the peers, which will fail the `newStream` inside Pubsub (as an example). Other than that, we will need to also try to connect with peers from the peerStore that are not connected

Reply:
This is only needed right now because Identify hasn't been refactored yet, right? The Registrar shouldn't need to care about multicodecs once identify is running. I think eventually we could have a peer discovery mechanism that checked the multicodecs our node supports (those added via `libp2p.handle`) and querying the network or rendezvous servers for nodes that also support those. This still wouldn't require getting multicodecs from the topology as we'll know them from `handle`/`unhandle` calls.

Reply:
I think this [conversation](https://github.com/libp2p/js-libp2p/pull/471#discussion_r342507203) has more insights. 

Basically, I am using them for getting the already connected peers.

I have a doubt though: Once pubsub starts, a `identify-push` will occur. If we have already connected and identified other peers (who support pubsub), we will already have that information right? This way, `protocols` will not change and we need to try to connect using pubsub on those peers.